ghost/core/core/server/services/members/members-api/controllers/RouterController.js

```diff
@@ -649,6 +651,85 @@ module.exports = class RouterController {
         }
     }
 
+    async verifyOTC(req, res) {
+        const {otc, otcRef} = req.body;
+
+        if (!otc || !otcRef) {
+            throw new errors.BadRequestError({
+                message: tpl(messages.badRequest),
+                context: 'otc and otcRef are required'
+            });
+        }
+
+        const tokenProvider = this._magicLinkService.tokenProvider;
+        // Validate OTC format using provider (fallback to 6 digits if unavailable)
+        const isFormatValid = tokenProvider?.isOTCFormatValid?.(otc) ?? /^\d{6}$/.test(otc);
+
+        if (!isFormatValid) {
+            throw new errors.BadRequestError({
+                message: tpl(messages.invalidCode)
+            });
+        }
+
+        try {
+            if (!tokenProvider || typeof tokenProvider.verifyOTC !== 'function') {
+                throw new errors.BadRequestError({
+                    message: 'OTC verification not supported'
+                });
+            }
+
+            const isValidOTC = await tokenProvider.verifyOTC(otcRef, otc);
+
+            if (!isValidOTC) {
+                res.writeHead(400, {'Content-Type': 'application/json'});
+                return res.end(JSON.stringify({
+                    valid: false,
+                    message: tpl(messages.invalidCode)
+                }));
+            }
+
+            const tokenValue = await tokenProvider.getTokenByRef(otcRef);
+            if (!tokenValue) {
+                res.writeHead(400, {'Content-Type': 'application/json'});
+                return res.end(JSON.stringify({
+                    valid: false,
+                    message: tpl(messages.invalidCode)
+                }));
```

Comment:
We should be following the newer pattern of `throw new errors.BadRequestError(...` rather than writing to the response directly. The direct `res.*` approach is a very old hangover from when `members-api` was a separate app from Ghost